syntax = "proto3";

message Header {
  message SockInit {
    uint32 id = 1;
    uint32 family = 2;
    uint32 type = 3;
    bytes preload_send = 4;
  }
  repeated SockInit socks = 1;
}

message Command {
  message SockSocket { uint32 family = 1; uint32 type = 2; uint32 proto = 3; uint32 target_id = 4; }
  message SockSocketpair { uint32 family = 1; uint32 type = 2; uint32 proto = 3; uint32 id1 = 4; uint32 id2 = 5; }
  message SockDup { uint32 src_id = 1; uint32 dst_id = 2; }
  message SockClose { uint32 id = 1; }
  message InetPton { uint32 family = 1; string text = 2; }
  message InetNtop { uint32 family = 1; bytes packed = 2; }
  message InetAton { string text = 1; }
  message InetNtoa { bytes packed = 1; }
  message Htons { uint32 val = 1; }
  message Htonl { uint32 val = 1; }
  message Ntohs { uint32 val = 1; }
  message Ntohl { uint32 val = 1; }
  message Getaddrinfo { string host = 1; string service = 2; uint32 family = 3; uint32 type = 4; uint32 proto = 5; uint32 flags = 6; }
  message Getnameinfo { bytes sockaddr = 1; uint32 flags = 2; }
  message IfNameindex {}
  message IfNametoindex { string name = 1; }
  message IfIndextoname { uint32 index = 1; }
  message Gethostbyname { string name = 1; }
  message GethostbynameEx { string name = 1; }
  message Gethostbyaddr { string addr = 1; }
  message Getservbyname { string name = 1; string proto = 2; }
  message Getservbyport { uint32 port = 1; string proto = 2; }
  message SockAccept { uint32 id = 1; uint32 new_id = 2; }
  message SockBind { uint32 id = 1; bytes addr = 2; }
  message SockListen { uint32 id = 1; uint32 backlog = 2; }
  message SockConnect { uint32 id = 1; bytes addr = 2; }
  message SockConnectEx { uint32 id = 1; bytes addr = 2; }
  message SockShutdown { uint32 id = 1; uint32 how = 2; }
  message SockSetsockopt { uint32 id = 1; uint32 level = 2; uint32 opt = 3; bytes val = 4; }
  message SockGetsockopt { uint32 id = 1; uint32 level = 2; uint32 opt = 3; }
  message SockSend { uint32 id = 1; bytes data = 2; uint32 flags = 3; }
  message SockSendto { uint32 id = 1; bytes data = 2; uint32 flags = 3; bytes addr = 4; }
  message SockSendall { uint32 id = 1; bytes data = 2; uint32 flags = 3; }
  message SockSendmsg { uint32 id = 1; bytes data = 2; uint32 flags = 3; }
  message SockRecv { uint32 id = 1; uint32 maxlen = 2; uint32 flags = 3; }
  message SockRecvfrom { uint32 id = 1; uint32 maxlen = 2; uint32 flags = 3; }
  message SockRecvmsg { uint32 id = 1; uint32 maxlen = 2; uint32 ancbufsize = 3; uint32 flags = 4; }
  message SockRecvInto { uint32 id = 1; uint32 nbytes = 2; uint32 flags = 3; }
  message SockRecvfromInto { uint32 id = 1; uint32 nbytes = 2; uint32 flags = 3; }
  message SockRecvmsgInto { uint32 id = 1; uint32 nbytes = 2; uint32 ancbufsize = 3; uint32 flags = 4; }
  message SockSettimeout { uint32 id = 1; double timeout = 2; }
  message SockSetblocking { uint32 id = 1; bool blocking = 2; }

  oneof cmd {
    SockSocket sock_socket = 1;
    SockSocketpair sock_socketpair = 2;
    SockDup sock_dup = 3;
    SockClose sock_close = 4;
    InetPton inet_pton = 5;
    InetNtop inet_ntop = 6;
    InetAton inet_aton = 7;
    InetNtoa inet_ntoa = 8;
    Htons htons = 9;
    Htonl htonl = 10;
    Ntohs ntohs = 11;
    Ntohl ntohl = 12;
    Getaddrinfo getaddrinfo = 13;
    Getnameinfo getnameinfo = 14;
    IfNameindex if_nameindex = 15;
    IfNametoindex if_nametoindex = 16;
    IfIndextoname if_indextoname = 17;
    Gethostbyname gethostbyname = 18;
    GethostbynameEx gethostbyname_ex = 19;
    Gethostbyaddr gethostbyaddr = 20;
    Getservbyname getservbyname = 21;
    Getservbyport getservbyport = 22;
    SockAccept sock_accept = 23;
    SockBind sock_bind = 24;
    SockListen sock_listen = 25;
    SockConnect sock_connect = 26;
    SockConnectEx sock_connect_ex = 27;
    SockShutdown sock_shutdown = 28;
    SockSetsockopt sock_setsockopt = 29;
    SockGetsockopt sock_getsockopt = 30;
    SockSend sock_send = 31;
    SockSendto sock_sendto = 32;
    SockSendall sock_sendall = 33;
    SockSendmsg sock_sendmsg = 34;
    SockRecv sock_recv = 35;
    SockRecvfrom sock_recvfrom = 36;
    SockRecvmsg sock_recvmsg = 37;
    SockRecvInto sock_recv_into = 38;
    SockRecvfromInto sock_recvfrom_into = 39;
    SockRecvmsgInto sock_recvmsg_into = 40;
    SockSettimeout sock_settimeout = 41;
    SockSetblocking sock_setblocking = 42;
  }
}

message Program {
  Header init = 1;
  repeated Command cmds = 2;
}

